### ISSUE TITLE
Support for security scheme/auth

### ISSUE THREAD

**Issue by czechboy0** [apple/swift-openapi-generator#37]
https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.3.md#security-scheme-object

Let's do a proposal first, as there are a few ways this could be implemented.

For now, the workaround is to write your own middleware that injects the appropriate headers to every outgoing request.

Example: https://github.com/apple/swift-openapi-generator/tree/main/Examples/auth-client-middleware-example

**Comment by Jonovono** [comment]
@czechboy0 I'm able to get the client middleware to inject headers to the request, but whats the best way to extract that on the server?

What I have tried so far is having a server middleware that verifies the token:

```
    func intercept(
        _ request: Request,
        metadata: ServerRequestMetadata,
        operationID: String,
        next: (Request, ServerRequestMetadata) async throws -> Response
    ) async throws -> Response {
        if protectedOperations.contains(operationID) {
            if let authToken = request.headerFields.first(where: { $0.name == "Authorization" })?.value
            {

                if validateJWTToken(authToken) {
                    var newHeaderFields = request.headerFields
                    newHeaderFields.append(HeaderField(name: "uid", value: "the-user-id"))

                    let newRequest = Request(
                        path: request.path,
                        query: request.query,
                        method: request.method,
                        headerFields: newHeaderFields,
                        body: request.body
                    )
                    let response = try await next(newRequest, metadata)
                    return response
```

All that works great, but from the auth jwt token I extract a uid, i'd like to pass that along to the subsequent handlers. How could I achieve this?

In the code above, I tried adding it to the headers, but then in my handler I don't seem to see it:

```
    func authorizeUser(_ input: Operations.authorizeUser.Input) async throws -> Operations.authorizeUser.Output {
        print("INPUT", input.headers)
        return .ok(.init())
    }

```

Any ideas? I might be doing it completely wrong

**Comment by czechboy0** [comment]
@Jonovono you could add the value to a TaskLocal value on the server, which allows you to access it from the handler. Or inject a token storage object both into the middleware and the object implementing APIProtocol on the server.

**Comment by Jonovono** [comment]
Thanks @czechboy0 , I have went the TaskLocal route and seems to be working good!

**Comment by mapedd** [comment]
Right now, for default handlers generated by OpenAPI for Vapor. there's no way to access request object to authorize it. Is there any solution that being worked on?

**Comment by czechboy0** [comment]
Yes, there are two solutions to this, one you can use now, and one tracked by this issue that aims to generate some type-safe representation of the authorization rules from the OpenAPI document.

The first solution is to create a Vapor `AsyncMiddleware` type and save whatever information you'd like to pass along into a task local value. Then in the handler (your type conforming to `APIProtocol`, you extract the task local value and use it).

Something like this should work:

```swift
import OpenAPIVapor
import Vapor
​
enum MyTaskLocal {
    @TaskLocal
    static var vaporValue: String? // Or whatever value you're extracting.
}
​
struct TaskLocalVaporMiddleware: AsyncMiddleware {
    func respond(
        to request: Vapor.Request,
        chainingTo next: Vapor.AsyncResponder
    ) async throws -> Vapor.Response {
        try await MyTaskLocal.$vaporValue.withValue(request.<EXTRACT_VALUE_HERE>) {
            try await next.respond(to: request)
        }
    }
}
​
struct Handler: APIProtocol {
    func test(_ input: Operations.test.Input) async throws -> Operations.test.Output {
        print("Task local from Vapor middleware: \(MyTaskLocal.vaporValue ?? "<nil>")")
        return .noContent(.init())
    }
}
​
@main
struct Tool {
    static func main() async throws {
        let app = Application()
        app.middleware.use(TaskLocalVaporMiddleware())
        let transport = VaporTransport(routesBuilder: app)
        let handler = Handler()
        try handler.registerHandlers(on: transport)
        try app.run()
    }
}
```

One caveat, the last time I tested this, the middleware had to be the last one in the chain (otherwise it doesn't work), so add it right before you start the server. This allows you to pass values from the concrete Vapor.Request type all the way to your type-safe handlers.

**Comment by mapedd** [comment]
This looks great, thanks!

**Comment by marius-se** [comment]
> Something like this should work:
...

What's the go-to way (until we get proper support) for registering different middlewares for different routes? My current solution is to switch over `request.route` in `TaskLocalVaporMiddleware.respond`, but that seems far from ideal.

**Comment by czechboy0** [comment]
The ServerMiddleware protocol provides an operationId parameter, which you can use to change behavior based on the operation.

Alternatively, you can inspect the path, e.g. if your auth requirements can be derived from just some common path prefix.

I'd have to learn more about your use case to make more precise recommendations.

**Comment by marius-se** [comment]
I essentially have two types of endpoints:
1. Public endpoints: Don't need any middleware
2. Private endpoints: Only authenticated requests should reach the request handler.

In Vapor we usually solve this using:
```
routes.get("overview", use: handler1)
routes.get("another-public-route": use: handler2)

// all routes registered under "private" will use AuthenticatorMiddleware
let private = routes.grouped(AuthenticatorMiddleware())
private.get("delete-me", use: handler3)
private.post("create-me", use: handler4)
```

And realistically there will be more endpoint-specific middlewares (e.g. some routes may require admin privileges etc.)

**Comment by czechboy0** [comment]
Gotcha. So if you don't want to have to maintain a manual list in your auth middleware that lets some operations through without auth and requires it for others, he's another idea.

Step 1: Use tags in your OpenAPI document to tag the operations that require authentication (e.g. called `authed`).
Step 2: Tag all the unauthed ones with a different one (e.g. called `unauthed`).
Step 3: Have 2 targets, e.g. called `AuthedRoutes` and `UnauthedRoutes`, and set up the generator for each of them, and in the config file of the first one, use the `filter` functionality to only include the `authed` and `unauthed` ones, respectively. Each would have its own `APIProtocol` protocol and its own handler implementation, e.g. called `AuthedHandler` and `UnauthedHandler`.
Step 4: In your server's executable, register them to the same server, such as:

```swift
import AuthedRoutes
import UnauthedRoutes

let app = Vapor.Application()

let transport = VaporTransport(routesBuilder: app)

// Register authed handler
let authedHandler = AuthedHandler()
try authedHandler.registerHandlers(on: transport, middlewares: [
  AuthenticationMiddleware()
])

// Register unauthed handler
let unauthedHandler = UnauthedHandler()
try authedHandler.registerHandlers(on: transport)

try app.run()
```

Could this work for you?

**Comment by marius-se** [comment]
@czechboy0 Thank you very much for your answer! While this does remove the additional route checking during runtime it is not very flexible. With each additional middleware this will require more symlinks and tags. Also it requires changing the OpenAPI yaml just to make something in Vapor work, which feels really wrong to me.

It is definitely a viable option for a small project, but for larger ones this will get out of control. I guess for now I'll stick to the additional path checking in the middleware!

Many thanks for you suggestions! 🙏🏼

**Comment by czechboy0** [comment]
Yeah it's not ideal, it's just a workaround for the lack of full support for security schemes in Swift OpenAPI Generator.

Once that's supported, you'd get all of the expected behavior almost for free, but it'll likely need to be contributed by someone who needs it, with our guidance. Could be a fun project 🙂

**Comment by jordanebelanger** [comment]
On my side I am using the URLSessionTransport for an iOS application, how can I pass in the bearer token to my API calls? The API I am using does not explicitly specify the `Authorization` header in its routes, but it does have the security field in the route for a bearer token security scheme. 

I tried subclassing URLSessionTransport to hijack the send method to inject my token only to find out that it's a struct so it cannot be subclassed.

Is there any easier way to do this? Surely I am missing something here 🤷

**Comment by czechboy0** [comment]
This example should be a good blueprint for what you need 🙂

https://github.com/apple/swift-openapi-generator/tree/main/Examples/auth-client-middleware-example

**Comment by czechboy0** [comment]
@tib yup, that's how it's expected to be used 👍

**Comment by daliad007** [comment]
Hello,

We're using `ClientMiddleware` to insert to correct auth info for the requests. However, when the user logs out / a different user logs in, those details change. To be able to set the new details on the client, the client object needs to be recreated as it doesn't seem to be a way to change the middleware once the client is instantiated. Is there a better way than this?

Thanks,
Dalia

**Comment by czechboy0** [comment]
Yes, you can make your middleware an `actor`, for example, or a `class` that you make `Sendable` and manually ensure safe access to internal mutable state. In this middleware stateful object, you can manage refreshing the credentials as needed, and always inject the current credentials in the `intercept` method.

That way, you can still create a single Client at the start, and reuse it even as the credentials change.

**Comment by edonv** [comment]
Any updates on this? I've actually made my own package to integrate like this. It doesn't do any codegen, but at least for now, it has protocols and whatnot to support different types of schemes, including an optional delegate that can be used to override the security scheme on a operation-by-operation basis. I don't have any real documentation yet (SPI should have some generated soon), but it works (I think lol). If it can be of any help, I'd be happy to help contribute to being part of the OpenAPIRuntime stuff.

https://github.com/edonv/swift-openapi-security-schemes

**Comment by simonjbeaumont** [comment]
@edonv thanks for sharing! ❤️ 

I haven't taken a look at it properly yet, but if you think you've landed on a generalisable shape here, which is sounds like is your aim, then we could look to explore adding it to the runtime package.

In general, although the runtime library only contains the middleware protocol definitions today, I'm open to us shipping actual middleware types in the runtime library, so long as they are suitably general purpose.

**Comment by czechboy0** [comment]
Agreed with Si here. Adding the building blocks to the runtime library would be a great step 1. Enhancing code generation to automatically inject (client) or enforce (server) auth based on the OpenAPI document could be a step 2. But we don't need to block (1) for (2), IMO.

If you'd be open to contributing that, writing a proposal would be a good first step, to make sure we can get everyone's input on the API shape.

**Comment by edonv** [comment]
Thanks for positive responses! The truth is that I've never written a proposal before and that kind of thing feels a little out of my depth. I've never really contributed to a large project before, plus it's not something I really have the time for. I'd be happy to be a resource, though, and help anyone work this into the runtime library and expand on my codebase as needed.

**Comment by simonjbeaumont** [comment]
@edonv no problem!

The proposal process might have sounded daunting, but it’s usually just a few hundred lines of markdown with some code examples of how it would be used. You can see them in the documentation bundle in the repo, if you wanted to see some examples. 

If this isn’t something you have the time or motivation to do, that’s completely OK. Thanks for taking the time to share the work you’ve done so far. We can keep this here and someone else might be able to drive it forward. 

Thanks again for your interest in the project! 🙏

**Comment by edonv** [comment]
Hey all, I'm back! So I still don't think I have the time to fully integrate this and write the proposal myself, but I'd love to at least help explain what I've written so it can be adapted as necessary into the Runtime library.

Here's how it works:
1. Each security scheme in an OpenAPI document (listed in `components/securitySchemes`) is defined as its own type (`struct` is fine).
    - Depending on the the `type` and `scheme` (per [OpenAPI's specs](https://spec.openapis.org/oas/latest.html#security-scheme-object)), it conforms to the appropriate protocol (which sub-protocol of [`SecurityScheme`](https://swiftpackageindex.com/edonv/swift-openapi-security-schemes/documentation/openapisecurityschemes/securityscheme) based on [`SecuritySchemeType`](https://swiftpackageindex.com/edonv/swift-openapi-security-schemes/documentation/openapisecurityschemes/securityschemetype)).
    - This part would be generated by the generator based on the document.
    - I'm seeing now that there's some overlap between my package and the Security section of [`OpenAPIKit`](https://github.com/mattpolzin/OpenAPIKit).
2. Once these types are set up, you create a `SecurityMiddleware` struct that conforms to [`SecuritySchemeMiddleware`](https://swiftpackageindex.com/edonv/swift-openapi-security-schemes/documentation/openapisecurityschemes/securityschememiddleware).
3. This type either needs to have a [`delegate`](https://swiftpackageindex.com/edonv/swift-openapi-security-schemes/documentation/openapisecurityschemes/securityschememiddleware/delegate) or a [`scheme`](https://swiftpackageindex.com/edonv/swift-openapi-security-schemes/documentation/openapisecurityschemes/securityschememiddleware/scheme-21xl3) property.
    - If there is only a single security scheme to apply to every operation, use `scheme`.
    - If the scheme needs to change on an operation-by-operation basis, use `delegate`.
        - The `delegate` (conforming to [`SecuritySchemeMiddlewareDelegate`](https://swiftpackageindex.com/edonv/swift-openapi-security-schemes/documentation/openapisecurityschemes/securityschememiddlewaredelegate)) has a delegate function that is used to return a `SecurityScheme`-conforming type (defined in step 1) based on the provided `operationId`.
    - If both a `delegate` and `scheme` are provided, the `scheme` is used when `delegate`'s function returns `nil`.
        - If different schemes are needed for different operations and some operations need *no* scheme, the best case is to use `delegate`, to not use `scheme`, and to return `nil` in `delegate`'s function when no scheme is needed.
4. Finally, just add an instance of the middleware (step 2) to your client (I add it to my class that wraps around my generated Client) or your server, and you're done.

Let me know what I can do to help going forward!

P.S. Some of the links are to the generated documentation on Swift Package Index. Some of those links' pages have more detail that hasn't been updated yet in the hosted documentation. It should be caught up sometime in the next 24 hours.

**Comment by czechboy0** [comment]
Hi @edonv,

thanks for sharing the work you've already done. I really like your holistic approach and covering both the client and the server.

Even if you don't have the time to integrate this, someone else might be able to pick up where you left off and turn this into a proposal.

**Comment by edonv** [comment]
I imagine the schemes themselves, the middleware, and the middleware delegate are all generated and/or abstracted in the Runtime library. The protocols, an abstracted middleware pre-attached to the client, and the bones of the typings would be in the Runtime library, and the concrete schemes from the spec document would be generated in the Generator.

**Comment by czechboy0** [comment]
Agreed, it seems much of this should go to the Runtime library, and we should only generate the Swift representation  of the knowledge of which operations use which auth coming from the OpenAPI doc.
